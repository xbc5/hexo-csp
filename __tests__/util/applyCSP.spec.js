"use strict";

const Config = require("../../lib/config");
const { applyCSP } = require("../../lib/util");
const template = require("../helpers/template");

// TODO:
// - test dev env;
//
describe("given valid args", () => {
  it("should return the markup with CSP applied", async () => {
    const c = {
      csp: {
        prod: {
          policies: {
            ["foo/index.html"]: {
              directives: {
                "default-src": ["'self'", "https://foo.com"],
                "img-src": ["https://bar.com"],
              },
            },
          },
        },
      },
    };
    const data = { path: "foo/index.html" };

    const config = new Config(c);
    const markup = applyCSP(config, data, template.noMeta);

    expect(markup).toMatchSnapshot();
  });
});

describe("given frontmatter", () => {
  it("should return the markup with CSP applied", async () => {
    const c = {
      csp: {
        prod: {
          policies: {
            ["foo/index.html"]: {
              directives: {
                "default-src": ["'self'", "https://foo-config.com"],
                "img-src": ["https://bar-config.com"],
              },
            },
          },
        },
      },
    };
    const data = {
      path: "foo/index.html",
      csp: {
        directives: {
          "default-src": ["https://foo-frontmatter.com"],
          "img-src": ["https://bar-frontmatter.com"],
        },
      },
    };

    const config = new Config(c);
    const markup = applyCSP(config, data, template.noMeta);

    expect(markup).toMatchSnapshot();
  });
});

describe("given no directives to apply", () => {
  [
    undefined,
    {},
    { csp: {} },
    { csp: { prod: {} } },
    { csp: { prod: { policies: {} } } },
  ].forEach((c) => {
    const param = c === undefined ? undefined : JSON.stringify(c);
    describe(`i.e. the config object is ${param}`, () => {
      it("it should return the original markup", async () => {
        const data = { path: "foo/index.html" };
        const config = new Config(c);
        const markup = applyCSP(config, data, template.noMeta);

        expect(markup).toBe(template.noMeta);
      });
    });
  });
});
