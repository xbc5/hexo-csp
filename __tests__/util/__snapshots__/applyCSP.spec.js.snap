// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given frontmatter should return the markup with CSP applied 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src 'self' https://foo-config.com; img-src https://bar-config.com"
    >
    <title>
      Fake Title
    </title>
  </head>
  <body>
  </body>
</html>
`;

exports[`given valid args should return the markup with CSP applied 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Security-Policy"
          content="default-src 'self' https://foo.com; img-src https://bar.com"
    >
    <title>
      Fake Title
    </title>
  </head>
  <body>
  </body>
</html>
`;
